### PR TITLE
Adding in the PubSub package.

### PR DESCRIPTION
**Pull request by ShriramKumar**

### CODE REVIEW

**Comment by yahoocla** [comment]
Thank you for submitting this pull request, however I do not see a valid CLA on file for you. Before we can merge this request please visit https://yahoocla.herokuapp.com/ and agree to the terms. Thanks! :smile:

File: pom.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
     <groupId>com.yahoo.bullet</groupId>
     <artifactId>bullet-core</artifactId>
-    <version>0.1.3-SNAPSHOT</version>
+    <version>0.2.0</version>
```
Comment (akshaisarma) [inline pom.xml]: It should be 0.2.0-SNAPSHOT. The release process strips it off when releasing.

File: src/main/java/com/yahoo/bullet/pubsub/Subscriber.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
+package com.yahoo.bullet.pubsub;
+
+public abstract class Subscriber {
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/pubsub/Subscriber.java]: Use interface and default methods

File: src/main/java/com/yahoo/bullet/pubsub/PubSubMessage.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import java.io.Serializable;
+import com.yahoo.bullet.pubsub.Metadata.Signal;
+import lombok.Getter;
+
+@Getter
+public class PubSubMessage implements Serializable {
+    private String id;
+    private String content;
+    private Metadata metadata;
+    private long sequenceNumber;
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/pubsub/PubSubMessage.java]: Use int, call it just sequence (it's clear it's a number) and can you move this to right after id?

File: src/main/java/com/yahoo/bullet/pubsub/PubSubMessage.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * Constructor for a PubSubMessage having content, Metadata and a sequence number.
+     *
+     * @param id is the query ID associated with the message.
+     * @param content is the content of the message.
+     * @param sequenceNumber is the sequence number associated with the message.
+     * @param metadata is the {@link Metadata} associated with the message.
+     */
+    public PubSubMessage(String id, String content, long sequenceNumber, Metadata metadata) {
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/pubsub/PubSubMessage.java]: All constructors should just call this one with the right arguments

File: src/main/java/com/yahoo/bullet/pubsub/PubSubMessage.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+public class PubSubMessage implements Serializable {
+    private String id;
+    private String content;
+    private Metadata metadata;
+    private long sequenceNumber;
+
+    /**
+     * Constructor for a message having only content.
+     *
+     * @param id is the query ID associated with the message.
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/pubsub/PubSubMessage.java]: We shouldn't mention that this is the queryID everywhere. I think we should just a class javadoc that mentions how PubSubMessage can be used with id being a query id and sequence being multi-part atomic messages per query.

File: src/test/java/com/yahoo/bullet/pubsub/PubSubMessageTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        Assert.assertNull(message.getMetadata().getContent());
+    }
+
+    @Test
+    public void testWithMetadataCreation() {
+        String messageId = getRandomString();
+        String messageContent = getRandomString();
+        String metadataContent = getRandomString();
+        Signal signal = Signal.ACKNOWLEDGE;
+        //Test creation with a sequence number.
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/pubsub/PubSubMessageTest.java]: Break it up into another test?

File: src/test/java/com/yahoo/bullet/pubsub/PubSubMessageTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        PubSubMessage message = new PubSubMessage(messageId, messageContent, 0, new Metadata(signal, metadataContent));
+        Assert.assertTrue(messageId.equals(message.getId()));
+        Assert.assertTrue(messageContent.equals(message.getContent()));
+        Assert.assertEquals(message.getSequenceNumber(), 0);
+        Assert.assertNotNull(message.getMetadata());
+        Assert.assertEquals(message.getMetadata().getSignal(), signal);
+        Assert.assertTrue(message.getMetadata().getContent().toString().equals(metadataContent));
+        //Test creation without a sequence number.
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/pubsub/PubSubMessageTest.java]: Another test

File: src/test/java/com/yahoo/bullet/pubsub/SubscriberTest.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import org.mockito.Mockito;
+import org.testng.annotations.Test;
+
+import java.util.UUID;
+
+public class SubscriberTest {
+    @Test
+    public void testCommitWithNoSequenceNumber() {
+        String randomID = UUID.randomUUID().toString();
+        Subscriber subscriber = Mockito.mock(Subscriber.class, Mockito.CALLS_REAL_METHODS);
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/pubsub/SubscriberTest.java]: If you test it, you don't have to mock it. Just create a minimal test implementation (can be in this file) and just call the default method if you need it.

**Comment by ShriramKumar** [comment]
Ready for re-review.

**Pull request merged**